cmd/updatetree/main_test.go

```diff
@@ -0,0 +1,178 @@
+			setFlags: resetFlags,
+			wantErr:  true,
+		},
+		{
+			desc: "validUpdateFrozen",
+			setFlags: func() {
+				*treeID = 12345
+				*treeState = "FROZEN"
+			},
+			wantRPC: true,
+			updateTree: &trillian.Tree{
+				TreeId:    12345,
+				TreeState: trillian.TreeState_FROZEN,
+			},
+			wantState: trillian.TreeState_FROZEN,
+		},
+		{
+			desc: "updateInvalidState",
+			setFlags: func() {
+				*treeID = 12345
+				*treeState = "ITSCOLDOUTSIDE"
+			},
+			wantErr: true,
+		},
+		{
+			desc: "unknownTree",
+			setFlags: func() {
+				*treeID = 123456
+				*treeState = "FROZEN"
+			},
+			wantErr:   true,
+			wantRPC:   true,
+			updateErr: errors.New("unknown tree id"),
+		},
+		{
+			desc: "emptyAddr",
+			setFlags: func() {
+				*adminServerAddr = ""
+				*treeID = 12345
+				*treeState = "FROZEN"
+			},
+			wantErr: true,
+		},
+		{
+			desc: "updateErr",
+			setFlags: func() {
+				*treeID = 12345
+				*treeState = "FROZEN"
+			},
+			wantRPC:   true,
+			updateErr: errors.New("update tree failed"),
+			wantErr:   true,
+		},
+	})
+}
+
+// runTest executes the updateTree command against a fake TrillianAdminServer
+// for each of the provided tests, and checks that the tree in the request is
+// as expected, or an expected error occurs.
+// Prior to each test case, it:
+// 1. Resets all flags to their original values.
+// 2. Sets the adminServerAddr flag to point to the fake server.
+// 3. Calls the test's setFlags func (if provided) to allow it to change flags specific to the test.
+func runTest(t *testing.T, tests []*testCase) {
+	for _, tc := range tests {
+		t.Run(tc.desc, func(t *testing.T) {
+			ctrl := gomock.NewController(t)
+			defer ctrl.Finish()
+
+			s, stopFakeServer, err := testonly.NewMockServer(ctrl)
+			if err != nil {
+				t.Fatalf("Error starting fake server: %v", err)
+			}
+			defer stopFakeServer()
+			defer flagsaver.Save().Restore()
+			*adminServerAddr = s.Addr
+			if tc.setFlags != nil {
+				tc.setFlags()
+			}
+
+			// We might not get as far as updating the tree on the admin server.
+			if tc.wantRPC {
+				call := s.Admin.EXPECT().UpdateTree(gomock.Any(), gomock.Any()).Return(tc.updateTree, tc.updateErr)
+				expectCalls(call, tc.updateErr)
+			}
+
+			ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second)
+			defer cancel()
+			tree, err := updateTree(ctx)
+			switch hasErr := err != nil; {
+			case hasErr != tc.wantErr:
+				t.Errorf("updateTree() returned err = '%v', wantErr = %v", err, tc.wantErr)
+				return
+			}
+
+			if err == nil {
+				if got, want := tree.TreeState.String(), tc.wantState.String(); got != want {
+					t.Errorf("updated state incorrect got: %v want: %v", got, want)
+				}
+			}
+		})
+	}
+}
+
+// expectCalls returns the minimum number of times a function is expected to be called
+// given the return error for the function (err), and all previous errors in the function's
+// code path.
+func expectCalls(call *gomock.Call, err error, prevErr ...error) *gomock.Call {
```

Comment:
I might not be following the Gomock stuff, but isn't only ever called with empty `prevErr` ?

Reply:
Yes, another case where it needs to do more work in creatree as there are several permuations of errors.